captain/models/topology.py

```diff
@@ -196,11 +196,13 @@ def process_job_result(
 
         logger.debug(f"out_edges to follow: {next_directions}")
 
-        for direction_ in next_directions:
+        for i, direction_ in enumerate(next_directions):
             direction = direction_.lower()
             if direction == "end" and self.loop_nodes:
                 self.loop_nodes.pop()
             self.mark_job_success(job_id, next_nodes_from_dependencies, direction)
+            if i == len(next_directions) - 1:
+                self.working_graph.remove_node(job_id)
```

Comment:
Maybe I am wrong, but wouldn't it be better to remove the edges out of which it flows instead of the entire node?

Reply:
This would create a conflict as there will be several nodes with out_degree == 0. So to be safe I considered removing node completely.